From 2e26f0e9bd27267fb2bc53c3dd5a1a437c085485 Mon Sep 17 00:00:00 2001
From: yevgeny.y.rouban <yevgeny.y.rouban@intel.com>
Date: Tue, 14 May 2013 15:53:42 +0700
Subject: Dalvik: fixed CompilationUnit changes lost by dvmCompilerMIR2LIR

BZ: 104876

The bug is in the lost CompilationUnit changes, which are needed by VTune jit dumper. The changes
are lost because dvmCompilerMIR2LIR sends an extended copy of the CompilationUnit to the backend and
the copy is not copied back to the original.

A quick fix would be to copy all the changes right before the dvmCompilerMIR2LIR exits (2 places).
But it seems a bit awkward to have 2 extra copying operations for every dvmCompilerMIR2LIR call.

The patch fixes dvmCompileTrace to instantiate CompilationUnit_O1 instead of CompilationUnit. Then,
dvmCompilerMIR2LIR just casts cUnit to CompilationUnit_O1*.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG; AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I3b9a7d87ef40310e94ac1fa3a293df39264fdff8
Orig-MCG-Change-Id: I3cc5d84026133752276dcc95615f16c003d3bada
Signed-off-by: yevgeny.y.rouban <yevgeny.y.rouban@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/Frontend.cpp                     |    8 ++++++++
 vm/compiler/codegen/x86/CodegenInterface.cpp |    6 +-----
 2 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/vm/compiler/Frontend.cpp b/vm/compiler/Frontend.cpp
index 6fd8dbf..5992e93 100644
--- a/vm/compiler/Frontend.cpp
+++ b/vm/compiler/Frontend.cpp
@@ -32,6 +32,10 @@
 extern int opcodeJit[kNumPackedOpcodes];
 #endif
 
+#ifdef ARCH_IA32
+#include "codegen/x86/CompilationUnit.h"
+#endif
+
 #if defined(VTUNE_DALVIK)
 #include "JitProfiling.h"
 #include <vector>
@@ -1982,7 +1986,11 @@ bool dvmCompileTrace(JitTraceDescription *desc, int numMaxInsts,
     BasicBlock *curBB, *entryCodeBB;
     int numBlocks = 0;
     static int compilationId;
+#ifndef ARCH_IA32
     CompilationUnit cUnit;
+#else
+    CompilationUnit_O1 cUnit;
+#endif
     GrowableList *blockList;
 #if defined(WITH_JIT_TUNING)
     CompilerMethodStats *methodStats;
diff --git a/vm/compiler/codegen/x86/CodegenInterface.cpp b/vm/compiler/codegen/x86/CodegenInterface.cpp
index c324738..ec6f4de 100644
--- a/vm/compiler/codegen/x86/CodegenInterface.cpp
+++ b/vm/compiler/codegen/x86/CodegenInterface.cpp
@@ -2380,11 +2380,7 @@ static void compilerMIR2LIRJit(CompilationUnit_O1 *cUnit, JitTranslationInfo *in
 void dvmCompilerMIR2LIR(CompilationUnit *cUnit, JitTranslationInfo *info) {
 
     //Make a x86 version of our CompilationUnit
-    CompilationUnit_O1 x86CUnit;
-
-    //Get a CompilationUnit of x86CUnit
-    CompilationUnit *child = &x86CUnit;
-    *child = *cUnit;
+    CompilationUnit_O1 &x86CUnit = *static_cast<CompilationUnit_O1*>(cUnit);
 
    //Save the optimization state to restore it at the end of this compilation
    SErrorCompilationState compilationState;
-- 
1.7.4.1

